### DISCUSSION TITLE
How to suspend a workflow if  there is an exception thrown and then  suspend the workflow, and then reume it with or without retry(by my choice) the previous failed activity?

### DISCUSSION THREAD

**Discussion by AngrySKL** [elsa-workflows/elsa-core#4993] [Q&A]
My scenario is that I have a long running worklfow with a lot of customized activities, some of the activities might throw exception when executing ( expected or unexpected ). And after the exception, I just want to suspend the workflow and then wait for the user to manually resume by clicking a "Resume" button. 

There are two kinds of suspension, one can be retried after resume and one cannot which means if the user click "Resume", if the activity caused the suspension can be retried then resume will retry the previous setp first and then going forward. If the activity can not be retried the resume will directly going forward.

I looked into the code of FaultStrategy, and maybe I can write a SuspensionStrastegy? But I don't see how to resume it...

**Comment by glime-ai** [comment]
To implement the behavior you're describing, you would indeed need to create a custom incident strategy that suspends the workflow instead of faulting it when an exception occurs. This strategy would be similar to the `FaultStrategy`, but instead of marking the workflow as faulted, it would suspend the workflow and potentially record some data that allows the user to resume the workflow later.

Here's a high-level approach to how you might implement this:

1. **Create a Custom Incident Strategy**: Implement an `IIncidentStrategy` that suspends the workflow. You can store information about the suspension in the workflow instance's data or as a custom entity in your database.

2. **Handling Suspension**: When an activity throws an exception, your custom incident strategy would be invoked, and you would handle the suspension logic there. You might want to add a flag or some metadata to indicate whether the activity can be retried or not.

3. **User Interface for Resuming**: In your application, you would provide a user interface that lists suspended workflows and allows users to click a "Resume" button.

4. **Resuming the Workflow**: When the user clicks "Resume", you would call the appropriate method to resume the workflow. If the activity can be retried, you would need to ensure that the workflow execution context is set up to retry the activity. If not, you would resume the workflow from the next activity.

5. **Implementing Resume Logic**: You might need to implement custom logic in your activities to handle the resume operation. This could involve overriding the `OnResumeAsync` method in your custom activities and adding logic to determine whether to retry the previous step or move forward.

6. **Workflow Instance Management**: Ensure that your workflow instances are persisted correctly and that you can query for suspended workflows. You might need to interact with the workflow instance store for this.

7. **Testing**: Thoroughly test your custom strategy and resume logic to ensure that workflows are suspended and resumed correctly, and that retryable activities are handled as expected.

Here's a pseudo-code example of what the custom incident strategy might look like:

```csharp
public class SuspendStrategy : IIncidentStrategy
{
    public void HandleIncident(ActivityExecutionContext context)
    {
        // Logic to suspend the workflow
        // You might want to add custom data to the workflow instance here
        // to indicate that it's suspended and whether it can be retried
    }
}
```

And to resume a workflow, you might have something like this:

```csharp
public async Task ResumeWorkflow(Guid workflowInstanceId, bool retry)
{
    var workflowInstance = await _workflowInstanceStore.GetByIdAsync(workflowInstanceId);
    if (workflowInstance == null)
        throw new InvalidOperationException("Workflow instance not found.");

    // Check if the workflow instance has the custom 'suspended' data
    // and whether the activity that caused the suspension can be retried

    if (retry)
    {
        // Logic to retry the activity
    }
    else
    {
        // Logic to move forward without retrying
    }

    // Resume the workflow
    await _workflowRunner.RunAsync(workflowInstance);
}
```

Remember to register your custom incident strategy with the DI container and configure it in your workflow options.

Please note that the actual implementation will depend on the specifics of your workflow and activities, and you may need to adjust the approach based on your requirements and the behavior of the Elsa Workflow engine.

**Comment by AngrySKL** [comment]
> To implement the behavior you're describing, you would indeed need to create a custom incident strategy that suspends the workflow instead of faulting it when an exception occurs. This strategy would be similar to the `FaultStrategy`, but instead of marking the workflow as faulted, it would suspend the workflow and potentially record some data that allows the user to resume the workflow later.
> 
> Here's a high-level approach to how you might implement this:
> 
> 1. **Create a Custom Incident Strategy**: Implement an `IIncidentStrategy` that suspends the workflow. You can store information about the suspension in the workflow instance's data or as a custom entity in your database.
> 2. **Handling Suspension**: When an activity throws an exception, your custom incident strategy would be invoked, and you would handle the suspension logic there. You might want to add a flag or some metadata to indicate whether the activity can be retried or not.
> 3. **User Interface for Resuming**: In your application, you would provide a user interface that lists suspended workflows and allows users to click a "Resume" button.
> 4. **Resuming the Workflow**: When the user clicks "Resume", you would call the appropriate method to resume the workflow. If the activity can be retried, you would need to ensure that the workflow execution context is set up to retry the activity. If not, you would resume the workflow from the next activity.
> 5. **Implementing Resume Logic**: You might need to implement custom logic in your activities to handle the resume operation. This could involve overriding the `OnResumeAsync` method in your custom activities and adding logic to determine whether to retry the previous step or move forward.
> 6. **Workflow Instance Management**: Ensure that your workflow instances are persisted correctly and that you can query for suspended workflows. You might need to interact with the workflow instance store for this.
> 7. **Testing**: Thoroughly test your custom strategy and resume logic to ensure that workflows are suspended and resumed correctly, and that retryable activities are handled as expected.
> 
> Here's a pseudo-code example of what the custom incident strategy might look like:
> 
> ```cs
> public class SuspendStrategy : IIncidentStrategy
> {
>     public void HandleIncident(ActivityExecutionContext context)
>     {
>         // Logic to suspend the workflow
>         // You might want to add custom data to the workflow instance here
>         // to indicate that it's suspended and whether it can be retried
>     }
> }
> ```
> 
> And to resume a workflow, you might have something like this:
> 
> ```cs
> public async Task ResumeWorkflow(Guid workflowInstanceId, bool retry)
> {
>     var workflowInstance = await _workflowInstanceStore.GetByIdAsync(workflowInstanceId);
>     if (workflowInstance == null)
>         throw new InvalidOperationException("Workflow instance not found.");
> 
>     // Check if the workflow instance has the custom 'suspended' data
>     // and whether the activity that caused the suspension can be retried
> 
>     if (retry)
>     {
>         // Logic to retry the activity
>     }
>     else
>     {
>         // Logic to move forward without retrying
>     }
> 
>     // Resume the workflow
>     await _workflowRunner.RunAsync(workflowInstance);
> }
> ```
> 
> Remember to register your custom incident strategy with the DI container and configure it in your workflow options.
> 
> Please note that the actual implementation will depend on the specifics of your workflow and activities, and you may need to adjust the approach based on your requirements and the behavior of the Elsa Workflow engine.

Can you show me how to retry the activity?

**Comment by sfmskywalker** [comment]
To retry a workflow instance that has not yet finished, but e.g. running with incidents (as in your case), you can use the [Alterations API](https://v3.elsaworkflows.io/docs/alterations/introduction) and specify the `ScheduleActivity` alteration. The documentation offers a sample request.

**Reply by AngrySKL** [reply]
Thanks a lot, I'll give it a try.

**Reply by AngrySKL** [reply]
Oh, meanwahile do I need to use bookmarks to suspend the workflow? If this is the case, I have to add an additional Bookmark Activity after each my customized activities? Is there a scheduled suspension or termination etc...?

**Reply by AngrySKL** [reply]
Or I can use the IncidentStrategy to create a bookmark, right?

**Reply by AngrySKL** [reply]
I create an activity to throw exception and an IncidentStrategy to create the bookmark , well the workflow is suspended after the exception  by the bookmark, but when I tried to get the bookmark and resume the workflow with it, an exception was thrown that tells me there is no bookmark at all
<img width="524" alt="image" src="https://github.com/elsa-workflows/elsa-core/assets/8011904/67bf763d-a08f-4fb5-9c9a-9fe0dc01b09b">

I simply modified the sample Elsa.Samples.ConsoleApp.Bookmarks to do so and below is the code, could help to take a look what's wrong? Thanks a lot @sfmskywalker 

[CreateBookmarkAfterException.zip](https://github.com/elsa-workflows/elsa-core/files/14395232/CreateBookmarkAfterException.zip)

**Reply by ErnestManukyanPlatAi** [reply]
@AngrySKL did you solve this ?

**Comment by glime-ai** [comment]
To implement the behavior you're describing, you would indeed need to create a custom incident strategy that handles exceptions by suspending the workflow instead of faulting it. This strategy would be responsible for creating a bookmark that can be used to resume the workflow later.

Here's a high-level approach to how you might implement this:

1. **Create a Custom Incident Strategy**: Implement an `IIncidentStrategy` that, instead of faulting the workflow, creates a bookmark to suspend the workflow. This bookmark could include metadata indicating whether the activity can be retried or not.

2. **Handling the Resume**: When the user clicks the "Resume" button, you would need to implement logic that checks the bookmark metadata to determine if the activity should be retried. If it should be retried, you would execute the activity again; if not, you would move on to the next activity in the workflow.

3. **Implementing the Resume Logic**: You would need to use the `IWorkflowRunner` or `IWorkflowHost` service to resume the workflow from the bookmark. This would involve retrieving the workflow instance and the bookmark, and then calling the appropriate method to resume the workflow with the bookmark.

4. **UI Integration**: On the user interface side, you would need to provide a way for users to trigger the resume operation. This would involve some form of button or action that, when clicked, calls the backend service to resume the workflow.

5. **Error Handling**: Ensure that your custom strategy and resume logic handle exceptions gracefully, logging any issues and providing feedback to the user if the resume operation cannot be completed.

6. **Testing**: Thoroughly test your custom strategy and resume logic to ensure that they work correctly with both retryable and non-retryable activities.

For the actual implementation, you would need to dive into the Elsa Workflow documentation and source code to understand how to create custom incident strategies and bookmarks, as well as how to interact with the workflow runner services to resume workflows.

Remember to also consider the security implications of allowing users to resume workflows, as you may need to implement authorization checks to ensure that only authorized users can resume certain workflows or activities.

Since the Elsa Workflow documentation and source code are your primary sources for how to implement these features, you should refer to them for detailed guidance on implementing your custom strategy and resume logic.
